### DISCUSSION TITLE
Improve typing

### DISCUSSION THREAD

**Discussion by karpetrosyan** [encode/httpcore#701] [Ideas]
Should we replace `Dict[str, any]` with `typing.TypedDict` as an Extensions type?

Now:

``` python
Extensions = Mapping[str, Any]
```

That means any dictionary with the key type of `str` can be passed, and the type checker will recognize it as a valid dictionary.

``` python
import httpcore
httpcore.Request(..., extensions={'not_exists': type}) # valid
```

Usage of TypeDict

``` python
import typing

class Extensions(typing.TypedDict):
    timeout: int

invalid_req: Extensions = {'timeout': type} # invalid (type)
valid_req: Extensions = {'timeout': 1} # valid
invalid_req: Extensions = {'t': 1} # invalid (missing timeout)
```

It's a good UX improvement, in my opinion.

**Comment by lovelydinosaur** [comment]
That's an interesting suggestion, and you're not the first to make it.

I'm not sure about it because...

The extensions that are valid depend on the transport that's in use. Using the dict typing here implies (or is implied by) that loose coupling.
